lib/charms/opensearch/v0/opensearch_backups.py

```diff
@@ -235,105 +258,193 @@ def _restore_and_try_close_indices_if_needed(
+            },
+        )
+        logger.debug(f"_issue_restore_request: restore call returned {output}")
+        if (
+            self.get_service_status(output)
+            == BackupServiceState.SNAPSHOT_RESTORE_ERROR_INDEX_NOT_CLOSED
+        ):
+            to_close = output["error"]["reason"].split("[")[2].split("]")[0]
+            raise OpenSearchRestoreFailedClosingIdxError(
+                f"_issue_restore_request: fails to close {to_close}"
+            )
+        return output
+
+    def _check_if_restore_finished(self) -> bool:  # noqa: C901
+        rel = self.model.get_relation(PeerRelationName)
+        if not rel:
+            # Running on a single unit, wait_for_completion=false not supported
+            return True
+        if not rel.data[self.charm.app].get("restore_in_progress"):
+            # Dealing with an empty set
+            return True
+        closed_idx = set(rel.data[self.charm.app].get("restore_in_progress", "").split(",")).copy()
+        # Check if all indices are open again
+        try:
+            indices_status = self._request(
+                "GET",
+                "/_recovery?human",
+            )
+        except (ValueError, OpenSearchHttpError, requests.HTTPError) as e:
+            raise OpenSearchRestoreCheckError(f"Failed GET to /_recovery with : {e}")
+        if isinstance(indices_status, str):
+            indices_status = json.loads(indices_status)
+        elif isinstance(indices_status, list) or isinstance(indices_status, set):
+            raise ValueError("Unexpected output from /_recovery: type list received")
+
+        indices_not_found = closed_idx.copy()
+        for idx, info in indices_status.items():
+            if idx in closed_idx:
+                indices_not_found.discard(idx)
+            else:
+                continue
+            # Now, check the status of each shard
+            done = True
+            for shard in info["shards"]:
+                if shard["type"] != "SNAPSHOT":
+                    continue
+                if shard["stage"] != "DONE":
+                    done = False
+                    break
+            if done:
+                closed_idx.discard(idx)
+
+        if indices_not_found:
+            raise OpenSearchRestoreMismatchBetweenIdxAndSnapshotError(
+                f"Failed to find indices {indices_not_found} in /_recovery"
+            )
+        if (
+            not closed_idx
+            and self.charm.unit.is_leader()
+            and "restore_in_progress" in rel.data[self.charm.app]
+        ):
+            del rel.data[self.charm.app]["restore_in_progress"]
+            return True
+        return False
+
+    def _on_check_restore_status_action(self, event: ActionEvent) -> None:
+        """Checks the status of indices that have been recovered from snapshot."""
+        try:
+            if self._check_if_restore_finished():
+                event.set_results({"state": "successful restore!"})
```

Comment:
before this you should probably add `self.charm.peers_data.delete(Scope.APP, "restore_in_progress")`